### ISSUE TITLE
Debugging a package

### ISSUE THREAD

**Issue by pat-s** [ManuelHentschel/VSCode-R-Debugger#56]
Hi Manuel,

thanks for the great work, a debugger for vscode is highly needed.

I am struggling to get it working for a package (haven't tried normal code yet).

Currently I am executing these steps:

1. Set a breakpoint in a function
1. Install the package with `devtools::install(keep_source = TRUE)`
1. Focusing an Rscript in the editor
1. Hitting F5 (debug console starts with message "Starting R session")
1. Executing some lines of code including a call to a function which contains the breakpoint

However, nothing happens and I do not know why. The R calls are executed in radian.

Probably I am missing something obvious here - highly appreciated if you could provide some guidance here :)

**Comment by ManuelHentschel** [comment]
Hi Patrick,

thanks for the feedback. The following points might help to narrow down the problem:
* Try debugging some normal code to check if the debugger itself is working
* Use vanilla R instead of radian (I haven't done any testing with radian yet)
* The message "Starting R session..." can be expanded to show some details about the debug session, could you post those please?

And which branch (master/develop) of the debugger are you using?

**Comment by renkun-ken** [comment]
I think debugging a package may not require that the package installed but `devtools::load_all()`. My use case to debug a package is that I set breakpoints in my testthat scripts and package source files.

When I run the debugging for a package, I expect that the `devtools::test()` is run. Is there a convenient way to specify the project settings to achieve this?

**Comment by pat-s** [comment]
@ManuelHentschel 

```
rDelimiter0: <v\s\c>
rDelimiter1: </v\s\c>
rPrompt: <#v\s\c>
rContinue: <##v\s\c>
rStartup: <v\s\c\R\STARTUP>
rLibraryNotFound: <v\s\c\LIBRARY\NOT\FOUND>
rAppend:  ### <v\s\c\COMMAND>
cwd: undefined
rPath: /usr/local/bin/R
rArgs: --ess --quiet --interactive --no-save
WARNING: unknown option '--ess'
> <v\s\c\R\STARTUP> 
library: vscDebugger
> > 
```

I just tried with a plain Rscript which contains 

```r
browser()
print("hurra")
```

but nothing is happening in the debugger (i.e. no variables, callstack, watch).
The debug console is initialized and when hitting F5 while focusing the R script then nothing happens.
Apparently I am doing something wrong - do you have an idea what it might be?

I am on macOS, `rdebugger.rterm` is valid.


@renkun-ken At least this is the default in RStudio, yes. Manuel specifically lists that one needs to use `devtools::install()` - I assume this is in purpose?

**Comment by ManuelHentschel** [comment]
Thanks for the info. You seem to be using a rather 'old' version of the debugger, so I would recommend reinstalling both the vscode extension and the R package from [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/tag/v0.1.0). Sorry for the bad versioning/dependency management so far, I'm trying to improve that by using Github releases.

Regarding the debugging of packages, the topic seems to be a bit more complicated. There is some info in the updated [readme](https://github.com/ManuelHentschel/VSCode-R-Debugger#debugging-r-packages). As long as the source information is preserverd during the installation, it should be possible to use `devtools::XXX()` or the commandline.

**Comment by pat-s** [comment]
Re-installing both helped, thanks!

Debugging scripts seems to work fine now. 

I can also debug R packages by doing the following:

1. Create a new R script and save it somewhere in the package.
1. ```
   pkgload::load_all()
   # function to debug which contains `browser()`
   ```
1. With focus on the file, either hit `F5` or click the "debug R-file" button


Here, 2) is a bit cumbersome as one needs to create an Rscript first, save it and then delete it again afterwards. Is it possible to call the debugger on an unsaved, temporary file somehow?

Also it would be great if an already existing R session could be used instead of spawning a new one by default. The vscode R ext attaches a default R session to each vscode session - maybe this one could be used if is found? This would remove the need of the `pkgload::load_all()` in the debug script.

Looks really promising so far! 🎉 I hope this project keeps growing - this is a very important initiative for the vscode-R usability :)

**Comment by ManuelHentschel** [comment]
I just tried debugging a package myself and realized that I recently changed some default settings without documenting the changes anywhere. In order to debug a package, the following two lines should be added to the launch config:
``` json
            "includePackageScopes": true,
            "setBreakpointsInPackages": true,
```
*(These are supposed to have different effects but apparently I messed up at some point, so just set both to `true` for now)*

Furthermore, debugging packages only seems to work with `"debugMode": "function"` at the moment, since this debug mode uses a different mechanism to set breakpoints.

Apparently, these issues only stem from errors in the option-handling/documentation, so I hope I'll be able to fix them rather quickly.


For reference, I attached a sample package and a screen recording showing how the debug process is supposed to work:
[RTestPackage.zip](https://github.com/ManuelHentschel/VSCode-R-Debugger/files/5069847/RTestPackage.zip)
[screenRecording.zip](https://github.com/ManuelHentschel/VSCode-R-Debugger/files/5069854/screenRecording.zip)

Does this example work for you?

**Comment by ManuelHentschel** [comment]
@renkun-ken 
> When I run the debugging for a package, I expect that the devtools::test() is run in the debugging session. Is there a convenient way to specify the project settings to achieve this?

To be honest, I don't really have much experience with tests in R. Could you maybe post a minimal example of a package with tests and annotate where/how you want breakpoints to work?

**Comment by pat-s** [comment]
@ManuelHentschel 

I am back after some time checking out your progress here. Seems like you are right on track, great! 🎉 

I've tried setting breakpoints and `browser()` statements in a package consisting only of R code.

I've tried all `launch.json` configurations adding 

```
            "includePackageScopes": true,
            "setBreakpointsInPackages": true,
```

to all configurations.


## Observations

They relate to the latest v0.4.1 of the debugger package.

- I am able to get to the constructor of the function which contains a breakpoint. I am not sure why it stops there however.
- When clicking on "continue", I am not sure what happens - the callstack and friends just get empty.
- Clicking on continue again, I again arrive at the constructor of the first function
- "file mode": Here I get the error "no function called `main` available`

```json
            "type": "R-Debugger",
            "request": "launch",
            "name": "Debug R-Function",
            "debugMode": "function",
            "includePackageScopes": true,
            "setBreakpointsInPackages": true,
            "workingDirectory": "${workspaceFolder}",
            "file": "${file}",
            // "mainFunction": "main",
            "allowGlobalDebugging": false
```

I was trying to follow your instructions for debugging packages in the README.
I think there is potential for optimizing the instructions - let me know if I can help (once I get it running in the first place).

Also if you want to have a live look at what I am doing and what I am possibly doing wrong, I would be happy to have a video call with you if you're up to :)
AFAIU you do not have the need for debugging packages so often - so if I can help out ;)
Judging from your name you are likely German so we could also do this in German then. 

(A functional debugger is the last bit that prevents me from switching completely to vscode so I would be up to help out wherever I can 👍)

**Comment by ManuelHentschel** [comment]
Welcome back 😉

First off, please make sure to install both the vscode extension and the R package from the latest master branch (I recently pushed some bug fixes that relate to debugging in packages). There are some tips how to do this in the [README](https://github.com/ManuelHentschel/VSCode-R-Debugger#installation), but let me know if you have any questions.

The required launch config to debug packages has changed (again) in the meantime:
* You need to specify the package(s) you want to debug as `"debuggedPackages": ["MyPackage"]`. This will set breakpoints in the package and try to overwrite `print`, `cat`, and `message`. Specifying a package here will also load it during the launch sequence, so you don't need to call `library("MyPackage")` in your scripts.
* `"setBreakpointsInPackages": true` is not recommended anymore, since it only looks in the exported functions, but does so for **all** packages (including the ones you don't want to debug)

Some more information about the launch config and other settings can be found in [configuration.md](https://github.com/ManuelHentschel/VSCode-R-Debugger/blob/master/configuration.md). Let me know if any of this helps to get the debugger running.

I would definitely appreciate your help with making the debugger easier to use and improving the documentation! And I'd also be up for a call to discuss this and/or have a look at your problems :)

**Comment by pat-s** [comment]
Thanks for the quick response!

I've followed the instructions again closely (I also did a few days ago already) and also tried setting `debuggedPackages` again. Unfortunately with no luck.

I am still stopping at the function definition instead of stopping at `browser()` or a defined breakpoint. I also cannot get to the latter by "continuing". I clearly must be doing something wrong. 

If you're up to, feel free to write me a mail (see mail in my profile) so we could arrange a meeting to sort this out and figure out how I can help you in this project :)
